### ISSUE TITLE
Add support for WOFF2

### ISSUE THREAD

**Issue by kennethmac2000** [opentypejs/opentype.js#183]
WOFF2 is now at Candidate Recommendation stage and is widely supported by the most popular browsers (including Chrome and Firefox).

Is it time to add support for it to opentype.js?

**Comment by fdb** [comment]
Yes.

Can you recommend a good Brotli library for JS?

**Comment by fdb** [comment]
For reference, the WOFF2 W3 doc: https://www.w3.org/TR/WOFF2/

**Comment by joseluis** [comment]
I believe only the decompressor is needed, right? These are the libs I know of:

1. https://github.com/devongovett/brotli.js/tree/master/dec/ decode.js
  - Instead of using dictionary-data.js use dictionary-browser.js and dictionary.bin. The total extra size may well be less than 100kb.
1. https://github.com/dominikhlbg/brotlijs ([discussion](https://groups.google.com/forum/#!topic/brotli/Y28CS-a7QHo))

**Comment by fdb** [comment]
Coo, I'll take a look.

Encoding would be cool as well as it would allow to write woff2 fonts (we don't do this for WOFF at the moment, but would like to support it there as well).

**Comment by devongovett** [comment]
I'm using my [brotli.js](https://github.com/devongovett/brotli.js) project in fontkit. Recently ported the decoder to JS by hand instead of using Emscripten. That ended up being much faster and smaller.

There's a bit more to decoding WOFF2 fonts than just brotli though. Here's the source code from fontkit: https://github.com/devongovett/fontkit/blob/master/src/WOFF2Font.coffee.

**Comment by fdb** [comment]
Thanks @devongovett! I know it's not just the decompression, but that's the part I didn't want to do myself :-)

**Comment by nazikus** [comment]
can someone provide a code sample how to decompress .woff2 file and parse it with opentype.js in browser?

**Comment by fdb** [comment]
Only woff is supported. Woff2 is not yet supported.

**Comment by caub** [comment]
this other repo seems to support woff2

**Comment by kontur** [comment]
From my superficial understanding woff2 is just a compression format. What in particular needs to get solved for this library to support woff2?

**Comment by Jolg42** [comment]
@kontur so the first official implementation of WOFF2 was not compatible to use from a browser (C++ & Node.js only).

But today thanks to [brotli.js](https://github.com/devongovett/brotli.js) this is now doable!

For me supporting WOFF2 is as easy as integrating brotli.js in the project, so feel free to contribute 😉

**Comment by kontur** [comment]
> For me supporting WOFF2 is as easy as integrating brotli.js in the project

In theory, for me, too :D

I sifted through the source to get a better understanding and checked the bramstein implementation, but I'm afraid this is way beyond my comprehension, no matter how much I'd like to help advance this. Pointers are welcome, of course, but I doubt I have the skills to implement this. From what I can tell `src/opentype.js` needs a different `parseWOFFTableEntries` specific to woff2 (different how?) and a different decompression in `uncompressTable` (using the brotli library linked, which I got to work, I think) based on the different signature detected in `parseBuffer` - but from my standpoint I'm mostly guessing at this. Is there still other aspects? :(

This library is included (second hand, via another library) in a plugin I author for type designers showcasing their webfonts. I've gotten requests to support woff2 in order to benefit from the improved load rates, so it would be great to see this feature added. If anybody can help, I and users of my plugin would really appreciate it.

**Comment by kontur** [comment]
Hey folks, I would really hope for this addition to come to the library. I've slowly been trying to get me head around how this library (and webfont internals) work and how to get woff2 to work, but I'm pretty lost.

In the branch referenced above I've tried to set up some basics needed for woff2 parsing (in the help those will be useful for others as a starting point), but in particular the parsing of the tables I don't fully comprehend, especially with those dynamic lengths of some of the fields and how to progress through them. I've added the brotli decompression, but since I'm lost with detecting the tables properly I can't really tell if it's going anywhere...

Anyway, this is a call for help, if anybody is more familiar with this library and finds this start to be of any help :)

**Comment by Jolg42** [comment]
I don't have time to check right now but I'll see if I can do something later!

**Comment by Jolg42** [comment]
I tried to debug your code, I think I'm almost there but I still have an issue...!

**Comment by kontur** [comment]
@Jolg42 Do you have it somewhere public so I could peek? I'd be curious (to educate myself ;)) what steps you took and what is still causing trouble.

**Comment by Jolg42** [comment]
@kontur only in local and there is a problem with the cmap table :(

**Comment by kontur** [comment]
Still desperately waiting for this feature <3

**Comment by kontur** [comment]
@Jolg42 Would you care to share you somewhat working version?

**Comment by kontur** [comment]
Is this anywhere near the upcoming releases even, anybody else interested in working together on this? :/

**Comment by Jolg42** [comment]
Hi @kontur 
So I didn't complete my first try, but I think it would be easier to use an existing project like
https://github.com/fontello/ttf2woff and https://github.com/nfroidure/ttf2woff2

**Comment by jlfwong** [comment]
Hi there, for a project I'm working on, I need to be able to read woff2 fonts in-browser.
I saw that sadly opentype.js doesn't yet have support for it, but I think I'll be able to work around it by converting the woff2 format to ttf first. That in itself is kind of a pain, but I think I have it working via an emscripten cross-compiled version of http://github.com/google/woff2. You can see my fork here: https://github.com/jlfwong/woff2-to-ttf.js

So my question is: would you accept a PR to add support for woff2 via this conversion? The huge downside is that it will significantly bloat the size of this library. The emscripten compiled source is around 730kB, and minification doesn't do much to reduce that.

**Comment by darkfrog26** [comment]
@jlfwong that seems like a really bad solution to the problem as apart from the increase in the source code, it will be incredibly inefficient to convert formats and then read the format rather than just read the format directly.  What I've had to do to work-around this problem is do server-side conversions instead from woff2 to ttf. While still not ideal, it at least doesn't increase the client's burden significantly, and the server can cache the conversions so it only has to it once per font.

**Comment by jlfwong** [comment]
@darkfrog26 The woff (woff1 and woff2) file formats are both fancy wrappers around sfnt fonts. Both Chrome & Firefox actually do confirm from woff to TTF before parsing the information, I believe. See https://www.w3.org/TR/WOFF2/#table_format and https://www.w3.org/TR/WOFF/#OverallStructure. Essentially, woff & woff2 aren't really font formats -- they're just per-font-table compression applied to existing font files. In order to read a woff file, you need to inflate the tables, then parse them with another method anyway.

**Comment by fdb** [comment]
Ideally there would be some system to turn them on/off as needed. Something like peer dependencies in npm perhaps (https://nodejs.org/en/blog/npm/peer-dependencies/)?

I would love for OpenType.js to work with @devongovett 's Brotli decompressor (https://github.com/devongovett/brotli.js/), which is hand-ported and generally awesome.

**Comment by Jolg42** [comment]
Like @fdb I think that brotli.js is the way to go! I tried to implement it a while ago but without success unfortunately... Didn't have time to finish it since neither.

**Comment by jlfwong** [comment]
Okay! I don't intend to help with that port, but look forward to that being done :) If anyone does have an immediate need to access woff2 fonts in-browser, https://github.com/jlfwong/woff2-to-ttf.js does indeed work now.

**Comment by wdanilo** [comment]
Oh, all google fonts are woff2 now. It is so bad we dont have support for it here :( By the way - is there any active maintainer here who is looking at this topic? :)

**Comment by darkfrog26** [comment]
What I did to handle that is to use TTF fonts from Google Fonts: https://github.com/outr/youi/blob/f02a4f24c784c0fb925e5b957e23f322ff9943f5/ui/shared/src/main/scala/io/youi/font/GoogleFont.scala

**Comment by jasonsantamaria** [comment]
Jumping in here to +1 support for WOFF2

**Comment by kontur** [comment]
Has anybody actually been able to use any of the preliminary testing I did in my April 2017 branch; as I recall I just tried to apply the decompression, but was stuck with the tables. I still would very much need this myself, but my programming experience is mostly in building interfaces, so with decompressing font tables I am somewhat lost. 
@Jolg42 if you have anything more than me, even with parts not working, I would love to peek and see what we can scratch together, even shared in private if you don't want "wip" code public.

It will soon turn 2 years since this issue was first opened, and by now woff2 is quickly overtaking woff in terms of actual webfont use. I wish I could do more :/

**Comment by vinuel** [comment]
It would be really cool to have a way to decompress WOFF2 files with [brotli.js](https://github.com/devongovett/brotli.js) first and then parse it with opentype.js ...

**Comment by shash7** [comment]
Any luck with this feature? I can chip in with a pull request if nobody's working on it.

**Comment by Jolg42** [comment]
@shash7 I will be happy to review it then!
I'm curious to know how you want to do it?

**Comment by shash7** [comment]
@Jolg42 I'm still going through the codebase trying to figure it all out. Seems like a fairly big undertaking at the moment.

I'll give you an update once I get woff2 parsing up and running.

**Comment by xqdoo00o** [comment]
> @Jolg42 I'm still going through the codebase trying to figure it all out. Seems like a fairly big undertaking at the moment.
> 
> I'll give you an update once I get woff2 parsing up and running.

@shash7  Hi，which brotli library would you use?

**Comment by shash7** [comment]
@Jolg42 @xqdoo00o Hey guys this is a bit beyond my capabilities at the moment. woff2 parsing is quite involved and I'm not sure to parse the woff2 tables inside parse.js

For future readers, there are a few libraries such as [fontkit](https://github.com/foliojs/fontkit), [opentype](https://github.com/bramstein/opentype), etc which can parse woff2.

**Comment by yne** [comment]
There are multiple WASM build of [google/woff2](https://github.com/google/woff2), for example https://github.com/fontello/wawoff2 which can be used to decompress/compress from/to a WOFF2 ArrayBuffer using 1 line of JS like `opentype.parse(Uint8Array.from(Module.decompress(buf)).buffer)`.

Here is a working sample to play with:

```html
<script src="https://unpkg.com/wawoff2@2.0.1/build/compress_binding.js"></script>
<form>
    <label>Compress:
        <input type=file accept=".ttf,.otf" onchange="files[0].arrayBuffer().then(buf=>form.size.innerText=Module.compress(buf).length)">
    </label>
    <label>
        WOFF2 Size:<output name=size>?</output>
    </label>
</form>
```

```html
<script src="https://unpkg.com/wawoff2@2.0.1/build/decompress_binding.js"></script>
<form>
    <label>Decompress:
        <input type=file accept=".woff2" onchange="files[0].arrayBuffer().then(buf=>form.size.innerText=Module.decompress(buf).length)">
    </label>
    <label>
        Original Size:<output name=size>?</output>
    </label>
</form>
```

Since the compress+decompress WOFF2 code takes around 900+300KB it's clearly a bad idea to bloat every OpenType.js build with 1.2MB of WOFF2-related compression code,
especially if it would have only required 1 external JS lib + 1 line of JS to do it "if needed".

But it's not the first nor the last issue raised on this topic so it would definitely be a good idea to document how to open/export WOFF2 from/to OpenType.js (on the README ?), so all WOFF2 related issues would be resolved.

**Comment by Connum** [comment]
Closed via #518
